sdk/src/main/java/com/mapbox/maps/renderer/FpsManager.kt

```diff
@@ -0,0 +1,186 @@
+        logI(TAG, "User defined ratio is $userDefinedRatio")
+      }
+    }
+  }
+
+  /**
+   * Return true if rendering should happen this frame and false otherwise.
+   *
+   * @param frameTimeNs time value from Choreographer in [System.nanoTime] timebase.
+   */
+  fun preRender(frameTimeNs: Long): Boolean {
+    // clear any scheduled task as new render call is about to happen
+    handler.removeCallbacksAndMessages(fpsManagerToken)
+    // no need to perform neither pacing nor FPS calculation when setMaxFps / setOnFpsChangedListener was not called by the user
+    if (userDefinedRatio == USER_DEFINED_RATIO_NOT_SET && fpsChangedListener == null) {
+      return true
+    }
+    preRenderTimeNs = System.nanoTime()
+    var skippedNow = 0
+    // check if we did miss VSYNC deadline meaning too much work was done in previous doFrame
+    if (previousRefreshNs != -1L && frameTimeNs - previousRefreshNs > screenRefreshPeriodNs + ONE_MILLISECOND_NS) {
+      skippedNow =
+        ((frameTimeNs - previousRefreshNs) / (screenRefreshPeriodNs + ONE_MILLISECOND_NS)).toInt()
+      choreographerSkips += skippedNow
+    }
+    previousRefreshNs = frameTimeNs
+    // we always increase choreographer tick by one + add number of skipped frames for consistent results
+    choreographerTicks += skippedNow + 1
+    if (skippedNow > 0 && LOG_STATISTICS) {
+      logW(
+        TAG,
+        "Skipped $skippedNow VSYNC pulses since last actual render," +
+          " total skipped in measurement period $choreographerSkips / $choreographerTicks"
+      )
+    }
+    if (userDefinedRatio != USER_DEFINED_RATIO_NOT_SET) {
+      return performPacing()
+    }
+    return true
+  }
+
+  fun postRender() {
+    fpsChangedListener?.let { listener ->
+      val frameRenderTimeNs = System.nanoTime() - preRenderTimeNs
+      frameRenderTimeAccumulatedNs += frameRenderTimeNs
+      // normally we update FPS counter once a second
+      if (choreographerTicks >= screenRefreshRate) {
+        calculateFps(listener)
+      } else {
+        // however to produce correct values we also update FPS after IDLE_TIMEOUT_MS
+        // otherwise when updating the map after it was IDLE first update will report
+        // huge delta between new frame and last frame (as we're using dirty rendering)
+        handler.postDelayed(IDLE_TIMEOUT_MS, fpsManagerToken) {
+          calculateFpsAndReset(listener)
+        }
+      }
+      preRenderTimeNs = -1L
+    }
+  }
+
+  fun onSurfaceDestroyed() {
+    fpsChangedListener?.let { listener ->
+      calculateFpsAndReset(listener)
+    }
+  }
+
+  fun destroy() {
+    handler.removeCallbacksAndMessages(fpsManagerToken)
+    fpsChangedListener = null
+  }
+
+  private fun calculateFpsAndReset(listener: OnFpsChangedListener) {
+    handler.removeCallbacksAndMessages(fpsManagerToken)
+    calculateFps(listener)
+    previousRefreshNs = -1
+  }
+
+  private fun performPacing(): Boolean {
```

Comment:
I would expect we could write tests for the every frame and simulate throttle / frame skips with low / high fps, right? Do you plan to write tests for the FpsManager itself?

Reply:
yeah, it is planned